src/logs/process.rs

```diff
@@ -11,34 +11,124 @@ use fancy_regex::Regex;
 use flate2::read::MultiGzDecoder;
 use itertools::Itertools;
 use lambda_runtime::Error;
+use std::collections::HashMap;
 use std::env;
 use std::ffi::OsStr;
 use std::io::Read;
 use std::ops::Range;
 use std::path::Path;
 use std::string::String;
 use std::time::Instant;
+use std::sync::{Arc, RwLock};
 use tracing::{debug, info};
 
 use crate::logs::config::{Config, IntegrationType};
 use crate::logs::coralogix;
 use crate::logs::ecr;
 
+
+pub struct MetadataContext {
+    inner: Arc<RwLock<HashMap<String, Option<String>>>>,
+}
+
+impl MetadataContext {
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub fn insert(&self, key: String, value: Option<String>) {
+        let mut inner = self.inner.write().unwrap();
+        inner.insert(key, value);
+    }
+
+    pub fn get(&self, key: &str) -> Option<String> {
+        let inner = self.inner.read().unwrap();
+        if let Some(v) = inner.get(key).cloned() {
+            v
+        } else {
+            None
+        }
+    }
+
+    /// evalueate dynamic metadata values for application_name and subsystem_name values.
+    /// values are expected to be in the format of<br>
+    /// `{{key}}` or `{{key|regex}}`, the regex must include a capture group which will be
+    /// used to extract the value. The key is the key in the metadata context map.
+    pub fn evaluate(&self, value: String) -> Result<String, String> {
+        if !value.starts_with("{{") && !value.ends_with("}}") {
+            return Ok(value);
+        };
+
+        let (reg, key) = if value.contains("|") {
+            let r = Regex::new(r#"\{\{\s?(?<key>[a-z\.0-9]+)\s?\|?\s?r'(?<regex>.*)'\s?\}\}"#)
```

Comment:
Recreating these regexes on each invocation of `evaluate` may be costly. I would recommend using something like `lazy_static` or `LazyLock` to create them once in a static variable.